components/browser/icons/src/main/java/mozilla/components/browser/icons/processor/ResizingProcessor.kt

```diff
@@ -0,0 +1,71 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.icons.processor
+
+import android.content.Context
+import android.graphics.Bitmap
+import androidx.annotation.VisibleForTesting
+import mozilla.components.browser.icons.Icon
+import mozilla.components.browser.icons.IconRequest
+import mozilla.components.support.ktx.android.content.res.pxToDp
+
+/**
+ * [IconProcessor] implementation for resizing the loaded icon based on the target size.
+ */
+class ResizingProcessor(
+    private val discardSmallIcons: Boolean = true
+) : IconProcessor {
+
+    override fun process(context: Context, request: IconRequest, resource: IconRequest.Resource?, icon: Icon): Icon {
+        val originalBitmap = icon.bitmap ?: return icon
+        val size = originalBitmap.width
+        val targetSize = context.resources.pxToDp(request.size.value)
+
+        // The bitmap has exactly the size we are looking for.
+        if (size == targetSize) return icon
+
+        val resizedBitmap = if (size > targetSize) {
+            resize(originalBitmap, targetSize)
+        } else {
+            // Our largest primary is smaller than the desired size. Upscale it (to a limit)!
+            // 'largestSize' now reflects the maximum size we can upscale to.
+            val largestSize = size * MAX_SCALE_FACTOR
+
+            if (largestSize > targetSize) {
+                // Perfect! WE can upscale and reach the needed size.
+                resize(originalBitmap, targetSize)
+            } else {
+                // We don't have enough information to make the target size look non-terrible.
+                // Best effort scale, unless we're told to throw away small icons.
+                if (discardSmallIcons) null else resize(originalBitmap, largestSize)
+            }
+        }
+
+        return icon.copy(bitmap = resizedBitmap).also { originalBitmap.recycle() }
+    }
+
+    /**
+     * Create a new bitmap scaled from an existing bitmap.
+     */
+    @VisibleForTesting
+    internal fun resize(bitmap: Bitmap, targetSize: Int) = try {
+        Bitmap.createScaledBitmap(bitmap, targetSize, targetSize, true)
+    } catch (e: OutOfMemoryError) {
+        // There's not enough memory to create a resized copy of the bitmap in memory. Let's just
+        // use what we have.
+        bitmap
+    }
+
+    companion object {
+        /**
+         * This is the largest factor we'll scale up an image by: the goal is an image
+         * that both fills the top site space well but does not have extreme resizing
+         * artifacts. This number was chosen anecdotally by comparing variously-sized
+         * favicons across devices to see which factor(s) looked the best. bug 1398970
+         * is filed to take a more comprehensive approach to favicons.
+         */
+        private const val MAX_SCALE_FACTOR = 3
```

Comment:
This introduces a new constant; although we already have `MAXIMUM_SCALE_FACTOR` in BrowserIcons which we pass to the decoders.

Reply:
I had to recompute `targetSize` as well. Should `process` be changed to take these variable as parameters, or alternatively should we add new fields to `IconRequest`? Could also just make `MAXIMUM_SCALE_FACTOR` internal instead of private.

Reply:
For now I'd just pass the to `process()` and not add them to the public API of `IconRequest`.